### PR TITLE
Mesh 1977/normalize legs

### PR DESCRIPTION
**Pull request by HenriqueNogara**
### modified API
- `LegAsset` has a new variant: `LegAsset::OffChain { ticker, amount }`; 
- `LegV2` has been renamed `Leg` and the old `Leg` no longer applies (i.e the types are different);
- The following variants have been removed from the Settlement error enum: `NoPendingAffirm`,  `LegNotPending`, `ReceiptNotClaimed`, `NoPortfolioProvided`, `LegCountTooSmall`,  `InstructionHasTooManyLegs`, `InvalidLegAsset`, `MaxNumberOfNFTsPerLegExceeded`, `DeprecatedCallOnV2Instruction`, `ReceiptForNonFungibleAsset`;
- The following variants have been added to the Settlement error enum: `MaxNumberOfFungibleAssetsExceeded`, `MaxNumberOfOffChainAssetsExceeded`, `NumberOfFungibleTransfersUnderestimated`, `UnexpectedOnChainAsset`, `UnexpectedOFFChainAsset`, `OffChainAssetCantBeLocked`, `OffChainAssetMustBeAffirmedWithReceipts`, `NumberOfOffChainTransfersUnderestimated`, `LegNotFound`;
- `InstructionLegsV2` storage has been renamed `InstructionLegs` and the old `InstructionLegs` no longer applies (i.e the types are different);
- The following extrinsics have been deleted: `add_instruction_with_memo`, `add_and_affirm_instruction_with_memo`, `add_instruction_with_memo_v2`, `add_and_affirm_instruction_with_memo_v2`, `affirm_instruction_v2`, `withdraw_affirmation_v2`, `reject_instruction_v2`;
- The following extrinsics have new parameters: `add_instruction`, `add_and_affirm_instruction`, `withdraw_affirmation`, `reject_instruction`, `affirm_with_receipts`, `affirm_instruction`;
- The `InstructionV2Created` event has been renamed `InstructionCreated` and the old event with the same name no longer applies (i.e the types have changed); 
### modified logic
- When adding a new instruction with `Fungible` or `NonFungible` asset, the ticker must exist on-chain.
- Receipts are only accepted for `OffChain` legs;
### data migration
- `InstructionLegsV2` data has been translated to fit the new `LegAsset`variant. All fungible assets that don't exist on chain are now of type `OffChain`; 
- The old  `InstructionLegs` data has been been translated to fit the new `Leg` type. All assets are either of type `Fungible`, if the assets can be found on-chain, or `OffChain` in case there are no assets on-chain,

### CODE REVIEW

File: pallets/settlement/src/lib.rs [hunk 1]
```diff
@@ -1527,21 +1408,17 @@ impl<T: Config> Module<T> {
                 Error::<T>::ReceiptAlreadyClaimed
             );
 
-            let leg = Self::get_instruction_leg(&id, &receipt.leg_id);
-            if let LegAsset::NonFungible(_nfts) = leg.asset {
-                return Err(Error::<T>::ReceiptForNonFungibleAsset.into());
-            }
+            let leg = InstructionLegs::get(&id, &receipt.leg_id).ok_or(Error::<T>::LegNotFound)?;
             ensure!(
```
Comment (HenriqueNogara) [inline pallets/settlement/src/lib.rs:1329]: The receiver would not be able to affirm if this check is not changed to this
```Rust
ensure!(
   portfolios_set.contains(&leg.from) || portfolios_set.contains(&leg.to),
   Error::<T>::PortfolioMismatch
);

**Review by adamdossa — APPROVED** [review]

File: pallets/settlement/src/lib.rs [hunk 2]
```diff
@@ -478,13 +380,25 @@ decl_module! {
...
+        #[weight = <T as Config>::WeightInfo::affirm_with_receipts(*fungible_transfers, *nfts_transfers, *offchain_transfers)]
+        pub fn affirm_with_receipts(
+            origin,
+            id: InstructionId,
+            receipt_details: Vec<ReceiptDetails<T::AccountId, T::OffChainSignature>>,
+            portfolios: Vec<PortfolioId>,
+            fungible_transfers: u32,
+            nfts_transfers: u32,
+            offchain_transfers: u32
+        ) -> DispatchResult {
```
Comment (Neopallium) [inline pallets/settlement/src/lib.rs]: Why does this need the 3 counts `*_transfers`?

File: pallets/settlement/src/lib.rs [hunk 3]
```diff
@@ -2016,50 +1882,139 @@ impl<T: Config> Module<T> {
...
+                        from: leg_v2.from,
+                        to: leg_v2.to,
+                        asset: LegAsset::NonFungible(nfts),
+                    },
+                }
+            };
+            InstructionLegs::insert(&id, leg_id, new_leg);
+        });
+        // Migrate all itens from InstructionLegs
+        v1::InstructionLegs::drain().for_each(|(id, leg_id, old_leg)| {
```
Comment (Neopallium) [inline pallets/settlement/src/lib.rs]: The order of the `InstructionLegsV2` and `InstructionLegs` migrations need to be reversed.

Right now the logic is moving the V2 legs to the same storage as the old V1 legs.  That will cause the `v1::InstructionLegs::drain()` to try returning both the old V1 and new V2 legs, which will fail to decode the V2 legs.

**Review by Neopallium — CHANGES_REQUESTED** [review]

File: pallets/settlement/src/lib.rs (hunk 2 above)
Comment (HenriqueNogara) [reply pallets/settlement/src/lib.rs]: this calls `lock_via_leg` which is influenced by the number of  `fungible` and `non-fungible` assets and `offchain` assets also have their own logic which is affected by their number  (e.g fetching the offchain legs and updating their status).

Comment (Neopallium) [reply pallets/settlement/src/lib.rs]: I am not a fan of all those count parameters.  We should look at making this work with just a leg count parameter, even if that means that some tx will cost more.

File: pallets/settlement/src/lib.rs [hunk 4]
```diff
@@ -1133,126 +548,132 @@ decl_module! {
...
 
         /// Provide affirmation to an existing instruction.
         ///
         /// # Arguments
-        /// * `id` - Instruction id to affirm.
+        /// * `id` - The `InstructionId` of the instruction to be affirmed.
         /// * `portfolios` - Portfolios that the sender controls and wants to affirm this instruction.
         /// * `fungible_transfers` - number of fungible transfers in the instruction.
         /// * `nfts_transfers` - total number of NFTs being transferred in the instruction.
+        /// * `weight_limit` - Optional value that defines a maximum weight for executing the instruction.
```
Comment (adamdossa) [inline pallets/settlement/src/lib.rs]: Comment needs to be updated?

File: pallets/settlement/src/benchmarking.rs [hunk 5]
```diff
@@ -925,78 +417,43 @@ benchmarks! {
...
+    }: _(alice.origin, InstructionId(1), receipt_details, portfolios, f, n, o)
     verify {
-        for (i, receipt) in receipt_details.iter().enumerate() {
-            assert_eq!(Module::<T>::instruction_leg_status(instruction_id, LegId(i as u64)),  LegStatus::ExecutionToBeSkipped(
-                receipt.signer.clone(),
-                receipt.receipt_uid,
-            ), "Settlement: Fail to affirm with receipts");
-        }
+        assert!(true);
     }
```
Comment (Neopallium) [inline pallets/settlement/src/benchmarking.rs]: We can just remove the `verify` block.

File: pallets/settlement/src/benchmarking.rs [hunk 6]
```diff
@@ -1006,168 +463,119 @@ benchmarks! {
         assert!(Module::<T>::receipts_used(&signer.account(), 0), "Settlement: change_receipt_validity didn't work");
     }
 
-    reschedule_instruction {
-        let l = T::MaxNumberOfFungibleAssets::get() as u32;
+    //reschedule_instruction {}: _()
+    //verify {}
```
Comment (Neopallium) [inline pallets/settlement/src/benchmarking.rs]: Don't we still need a benchmark for `reschedule_instruction`?

File: pallets/settlement/src/lib.rs [hunk 7]
```diff
@@ -842,55 +716,58 @@ decl_module! {
...
-                }
-            })
+            Ok(Self::base_execute_scheduled_instruction(id, &mut weight_meter))
         }
 
         /// Root callable extrinsic, used as an internal call to execute a scheduled settlement instruction.
-        #[weight = (*weight_limit).max(<T as Config>::WeightInfo::execute_scheduled_instruction(0, 0))]
-        fn execute_scheduled_instruction_v3(origin, id: InstructionId, weight_limit: Weight) -> DispatchResultWithPostInfo {
+        #[weight = (*weight_limit).max(<T as Config>::WeightInfo::execute_scheduled_instruction(0, 0, 0))]
+        fn execute_scheduled_instruction_v3(
```
Comment (Neopallium) [inline pallets/settlement/src/lib.rs:717]: We can change the names of these `execute_schedule_instruction*` extrinsics, just not change the order or parameters of the old ones.

But we can do that cleanup in a different PR.  I created MESH-2047 for adding a migration to remove the old schedule extrinsics.

File: pallets/settlement/src/benchmarking.rs (hunk 6 above)
Comment (HenriqueNogara) [reply pallets/settlement/src/benchmarking.rs]: I added it back. This should take the number of legs as a parameter, but I'll address this in the reschedule issue

**Review by adamdossa — APPROVED** [review]

File: pallets/settlement/src/lib.rs (hunk 1 above)
Comment (Neopallium) [reply pallets/settlement/src/lib.rs:1329]: The receiver shouldn't be using receipts to affirm their side.  Only the sender can use a receipt to affirm.

**Review by Neopallium — APPROVED** [review]
